ENH Use FieldValidators for FormField validation

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-framework/issues/11422

emteknetnz commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -2,29 +2,113 @@
+    /**
+     * Converter to convert date strings to another format for display in error messages
+     *
+     * @var callable
```
Not using strong typing because you cannot strongly type a callable property in PHP, even though you can assign it to an untyped property

emteknetnz commented on tests/php/Forms/ListboxFieldTest.php:
```diff
@@ -173,74 +173,63 @@ public function testFieldRenderingMultipleOn()
-            'Field validates values in source map'
-        );
-
-        $field->setValue(
```
This setValue() / assertion did nothing, the ArrayData is ignored in setValue(), so all it's testing is that setting a value of `false` is valid

emteknetnz commented on src/Forms/ListboxField.php:
```diff
@@ -256,7 +268,7 @@ public function getValueArray()
                 foreach ($value as $item) {
                     if (!is_array($item)) {
-                        $item = json_decode($item, true);
+                        $item = json_decode($item ?? '', true);
```
?? '' to prevent deprecation warning if setting a value of null

emteknetnz commented on src/Forms/TimeField.php:
```diff
@@ -420,12 +382,11 @@ protected function frontendToInternal($time)
+     * @return null|string The formatted time, or null if not a valid time
      */
-    protected function internalToFrontend($time)
+    public function internalToFrontend(mixed $value): ?string
```
Changed to public as required by FieldValidatorConverterInterface

emteknetnz commented on src/ORM/FieldType/DBField.php:
```diff
@@ -102,8 +102,6 @@ abstract class DBField extends ModelData implements DBIndexable, FieldValidation
         'ProcessedRAW' => 'HTMLFragment',
     ];
 
-    private static array $field_validators = [];
```
Duplicate as is already defined on FieldValidationTrait

emteknetnz commented on src/Forms/MultiSelectField.php:
```diff
@@ -72,11 +78,28 @@ public function setValue($value, $obj = null)
             $this->loadFrom($obj);
         } else {
-            parent::setValue($value);
+            parent::setValue($value, $obj);
```
This argument should have been passed through to parent before this PR

emteknetnz commented on src/Forms/FormField.php:
```diff
@@ -482,16 +511,6 @@ public function saveInto(DataObjectInterface $record)
-     * @see Formfield::setValue()
-     * @return mixed
-     */
-    public function dataValue()
```
Moved location in the file to be closer to the other *Value() methods

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -1221,30 +1240,30 @@ public function Type()
-     * @return bool
+     * Determines whether the field is valid or not based on its value
      */
-    protected function extendValidationResult(bool $result, Validator $validator): bool
```
Why are we getting rid of this? Did you look at the PR that added it and the reasoning for it? https://github.com/silverstripe/silverstripe-framework/pull/10569#discussion_r1013965980

emteknetnz replied:
Yes I did look at it. That's no longer compatible with the new way of doing this

The old `FormField::validate()` method had a signature of `validate(Validator $validator): bool`, whereas this has now been updated to match DataObject and DBField with `validate(): ValidationResult`

The replicate the extension functionality, there's a new extension hook `updateValidate` on FormField::validate() and a `ValidationResult` param to use. I've used this in combination with `$this->beforeExtending('updateValidate')` and calling parent::validate() on any overridden validate method so that the extension hook is always called.  I've added a unit test on `FormFieldTest::testUpdateValidateHookCalled()` to ensure this is happening

GuySartorelli replied:
Sweet. I think so long as that test is still there that's the main thing.

I do like that this now means the `updateValidate` extension hook is the same method signature for both `DataObject` and `FormField` - we should call that out in the changelog probably.

emteknetnz replied:
Yup already in the 6.0.0 changelog - https://github.com/silverstripe/developer-docs/pull/621/files#diff-027bda823f29903fb90f1ca3e86dc918c3b6a814c50096c66a3074b1da24eb62R797

emteknetnz replied:
Have mentioned in 6.0.0 changelog - https://github.com/silverstripe/developer-docs/pull/621/files#diff-027bda823f29903fb90f1ca3e86dc918c3b6a814c50096c66a3074b1da24eb62R799

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -444,16 +448,41 @@ public function getInputType()
-     * Returns the field value.
+     * Returns the internal value of the field
+     */
+    public function getValue(): mixed
```
We now have _four different variations_ on `getValue()`.... that's confusing as hell. Why do we need 4 different ways to get the value?

emteknetnz replied:
We need them all?

`getValue()` = Get the interval value of the field (this->value), which should be as "unmodified" as possible. Required to meet FieldValidationInterface
`getValueForValidation()` = Used for FieldValidators. Defaults to $this->getValue(). Something need modify this for some FieldValidators, such as CompositeFieldValidator() which validates child fields rather than the composite field itself. Required to the meet FieldValidationInterface.
`dataValue()` = Get a value the for insertion in the database. Will do some casting. If were were to get rid of one of the methods, I'd say it should be this one. But definitely not in this PR.
`Value()` = Get a value for use in templates. May modify the value quite a bit

Looking at this again, `getValueForValidation()` is possibly badly named as it's only intended to be used by FieldValidators, and not custom validate() methods which should use getValue() instead. There's a few options
a) Leave as is
b) Rename the method to `getValueForFieldValidation()`
c) Better doc block description
d) Remove from FormField + DBField and put it on FieldValidationTrait since they're identical

Perhaps b), c) and d)?

GuySartorelli replied:
Yeah, c/d sounds good.
In my mind the name you've suggested there is functionally identical so I don't care if it gets renamed to that or not. This is a field, so all of its validation is "field validation" even if it's not all using a FieldValidator.

That still doesn't solve the problem that we're going from two methods to four in this PR. Most developers will be able to rule out `getValueForValidation()` for their use case, but this PR still makes the decision go from "do I use `value()` or `dataValue()` to now `do I use `value()` or `getValue()` or `dataValue()`?


`getValue()` and `Value()` are so identically named that I would expect them to always return the same value, with one being a legacy alias for the other. 

> `Value()` = Get a value for use in templates. May modify the value quite a bit

I assume this is used via `getDefaultAttributes()` and `getSchemaStateDefaults()`, not directly via `$Value` in templates? If that's the case, it would be sensible to either rename it to something like `getValueForHtmlAttributes()` or else `transformValueForDataAttribute($value)` or something like that.
That makes it immediately clear that if I'm trying to just _get the value from the field in my PHP code_, I don't want to call that one.

That still leaves us with `getValue()` vs `dataValue()`, but at least that confusion already existing before this PR so I'd be okay with leaving it at that point.

emteknetnz replied:
> I assume this is used via getDefaultAttributes() and getSchemaStateDefaults(), not directly via $Value in templates?

I'm not sure where or how it's exactly used, my best understanding is that it prepares the value for display to a human. Here's a few examples of how it's implemented in subclasses

```php
class DateField extends TextField
// ..
    public function Value()
    {
        return $this->internalToFrontend($this->value);
    }
```

```php
class SiteTreeURLSegmentField extends TextField
// ...
    public function Value()
    {
        return rawurldecode($this->value ?? '');
    }
```

```php
class ReadonlyField
// ...
    public function Value()
    {
        // Get raw value
        $value = $this->dataValue();
        if ($value) {
            return $value;
        }

        // "none" text
        $label = _t('SilverStripe\\Forms\\FormField.NONE', 'none');
        return "<i>('{$label}')</i>";
    }
```

We can't use getValue() as an alias for Value(), as Value() isn't simply returning $this->value

Feels like the best we can do here is to have better PHP doc for both Value() and getValue()

GuySartorelli replied:
> We can't use getValue() as an alias for Value()

I'm not suggesting we do that at all - I'm saying that developers will expect that because they're basically the same named method. The name of the methods is what's problematic here.

I'm suggesting we rename `Value()` to something more specific so that developers don't have any confusion about what the difference between `getValue()` and `Value()` is. PHPDoc will help. Well named methods is better.

> I'm not sure where or how it's exactly used

Please find out where and how it's used, and then give it a name that is appropriate for that usage.
That will make it easier both to know when to call it, and to know what to implement inside that method when implementing new form fields.

emteknetnz replied:
In PHP It's primarily called within FormFields as $this->Value()
There's a handful of other places, seems to be mainly GridField related such as GridFieldNestedForm, GridFieldAddNewInlineButton that will call something like $formField->Value(). You could make an argument that alot of these should be switched to use getValue() instead, but that's outside the scope of this PR

It's used a lot in FormField templates as $Value

I think we should simply keep things as they are

GuySartorelli replied:
Keeping things as they are is not an option IMO. We can't add a `getValue()` and not rename `Value()` because the two methods are named too-similarly and it will cause unnecessary confusion. I'm already confused by them and this hasn't even been merged.

emteknetnz replied:
I've split off as [new card](https://github.com/silverstripe/silverstripe-framework/issues/11468), the scope of work is too large (~50 files to change) and will just muddy up this PR. Also a high risk of breaking stuff

GuySartorelli replied:
This PR adds the confusion, so this PR should clear it up, unless you can promise me the split off card is the very next thing you work on.

emteknetnz replied:
Yes happy to work on that after this is merged

GuySartorelli commented on src/Forms/DateField.php:
```diff
@@ -96,14 +106,6 @@ class DateField extends TextField
-     *
-     * @var mixed
-     */
-    protected $rawValue = null;
```
While I am glad to see we're not storing two sets of the value now... please explain what this was used for and why we don't need it anymore?

emteknetnz replied:
On DateField/DatetimeField/TimeField it was was an entirely unmodified value that came in from setValue() / setSubmittedValue(). Then the internal value $this->value was changed due to all the localisation fun that happens.  rawValue was then checked to see if the field was empty all along and to not validate it if was.

tl;dr it was pointless

GuySartorelli replied:
Sweet, sounds like it was a non-standard workaround for the way the value property was being (ab)used. I'm okay with it being axed.

GuySartorelli reviewed (changes requested):
Didn't do a full review, just looked at things that are relevant or adjacent to relevancy for the CMS 5 PRs.

GuySartorelli reviewed (changes requested):
Just giving this a quick test prior to looking at the code. I was using the `BasicFieldsTestPage` from silverstripe/frameworktest - note I had to remove the disabled dropdown fields, but I found that to be true without this PR as well.

1. dropdown fields message has changed from "Please select a value within the list provided. 0 is not a valid option" to the much less helpful "Not an allowed value"
1. With the `BasicFieldsTestPage` from silverstripe/frameworktest, I get "must be a string" validation errors for... I think it's the `FieldGroup`s? That doesn't happen before this PR
![image](https://github.com/user-attachments/assets/6129a31d-a2b5-43af-8a51-f3fca62027f9)
1. Setting random text in a `CurrencyField` doesn't give a validation error, it just pretends I didn't enter anything and clears the field after saving

emteknetnz commented:
> 1.

Yes some of the field messages have been diluted in the switch from custom messages in FormField::validate(), to the generic messages in FieldValidators.  In the case the message comes from OptionFieldValidator::getMessage().

In this particular circumstance, I think that's actually an OK message because a content author should never get into a situation where they can select an invalid option from a dropdown, you basically need malicious behaviour to trigger it, or it's a developer error.

Overall I think the diluted messages, where we don't show either the name of the field, or the invalid value, are a slight improvement if anything because in the CMS those message always show in context where they should attach themselves to the field that's invalid and the invalid value should still be in the field.  Note that we're consistently using ValidationResult $result->addFieldError(), rather than the form level $result->addError().

> 2.

Looks like there was couple of things wrong with frameworktest:
- Using Varchar columns for fields that are dealing with DataList's where ID is a numeric column - have changed their column types to Int instead
- Was setting the value of the disabled fields using DataObject->getField() which will return the database value of 0, instead of the correct value for the field which is null, which will skip FieldValidation.

You can see both of these updates in the force push comparison [here](https://github.com/silverstripe/silverstripe-frameworktest/compare/d481ac23faa47f14df9320471bebc57352f3ca70..394b9f3a4a15e2443f2b7d79f4d27614db9544b0)

> 3.

CurrencyField has existing logic in setValue() - `preg_replace('/[^0-9.\-]/', '', $value ?? '')`

GuySartorelli commented:
> Yes some of the field messages have been diluted in the switch from custom messages in FormField::validate(), to the generic messages in FieldValidators.

I don't think that's great - it definitely wasn't expected. Can you please list any other things I should expect as changes to the UX? From my perspective this was intended to essentially just change where the validation logic for formfields is. I wasn't expecting any changes to UX at all.

> Looks like there was couple of things wrong with frameworktest:
> - Using Varchar columns for fields that are dealing with DataList's where ID is a numeric column - have changed their column types to Int instead

You should be able to use a `Varchar` with dropdowns of any sort. An ID can be stored as a string, if that's what someone really wants to do, for whatever purposes they may have. 

Even if it we conclude you're not supposed to do that and it's a developer mistake (which I don't think we _can_ conclude), the result shouldn't be a confusing validation error that the content authors are confronted with.

> CurrencyField has existing logic in setValue() 

That's fair enough. Are we saying that changing/adding validation is out of scope - i.e. we don't really care if form fields aren't validated in a sensible way so long as it's the same as it was before?
If that's the case I'm happy to review to that expectation.

emteknetnz commented:
> I don't think that's great - it definitely wasn't expected. Can you please list any other things I should expect as changes to the UX?

I think it should be limited to validation messages now being different where FieldValidators are now being used. In the PRs I've tried to remove as much subclass FormField::validate() logic as I could, though I actually didn't get as far as I was expecting and there's still quite a bit of custom logic that I needed to retain as there's no FieldValidator equivalent.

> You should be able to use a Varchar with dropdowns of any sort. An ID can be stored as a string, if that's what someone really wants to do, for whatever purposes they may have.

My understanding here is that the issue isn't storing an ID as a string, the issue is that the value itself is an `int`, and you cannot save an `int` in a `DBVarchar` as that will fail validation. The ValidationException will be prior to this PR and will have started in the DBField PR instead, it's just this code path probably wasn't tested.

Validation messages from DBField will bubble through to the frontend components as FieldValidator will call ValidationResult `$result->addFieldError($this->name, ...` where $this->name is the same for both the DBField and the FormField.

My understanding is that thrown DBField ValidationException messages will be caught by FormRequestHandler

```php
        } catch (ValidationException $e) {
            // The ValidationResult contains all the relevant metadata
            $result = $e->getResult();
            $this->form->loadMessagesFrom($result);
            return $this->getValidationErrorResponse($result);
        }
```

Those particular dropdown where the I changed the DB column from Varchar to Int were using a DataList::map() as the source values, where the TestCategory->ID is the value.

The reason the submitted numeric string e.g. "3" is being cast to an int e.g. `3` is because of this newly added code in SingleSelectField (parent class of DropdownField)

```php
protected function getAllSourceValuesAreIntType(): bool
{
    $sourceValues = $this->getSourceValues();
    $intCount = count(array_filter(array_map('is_int', $sourceValues)));
    return $intCount === count($sourceValues);
}

public function setSubmittedValue($value, $data = null)
{
    if (is_numeric($value) && is_string($value) && $this->getAllSourceValuesAreIntType()) {
        $value = (int) $value;
    // ...
```

> Are we saying that changing/adding validation is out of scope - i.e. we don't really care if form fields aren't validated in a sensible way so long as it's the same as it was before?

No, the existing logic I left unchanged on CurrencyField is setValue(), which is not the same thing as validation.

In this PR I have added a few implementations of setSubmittedValue() in order to be smarter with casting values to the correct type. I have left setValue() alone though. Note that the base implementation of FormField::setSubmittedValue() will call FormField::setValue().

What this PR aims to do is replace custom custom validate() logic with standardised FieldValidation logic where possible, and along the way it will have also added additional fairly basic validation, such as checking string lengths of all textfields and that enum style fields e.g. Dropdowns are only allowed to select from a set of values. You can see a map of FormFields => FieldValidator's on [FormFieldTest::testFieldValidatorConfig()](https://github.com/silverstripe/silverstripe-framework/pull/11449/files#diff-1e0d5d4851ec39106d645d8e612091443a44afa553f873e621162c1007e6dfc7R766)

GuySartorelli commented:
> My understanding here is that the issue isn't storing an ID as a string, the issue is that the value itself is an int, and you cannot save an int in a DBVarchar as that will fail validation. 

Forms are untyped, realistically.  I think either `DBVarchar` should convert ints to string (the same way `DBInt` converts strings to int)  or this needs to be handled earlier than that - but it shouldn't result in validation errors.

I would say this would be part of https://github.com/silverstripe/silverstripe-framework/issues/11453 but that implies it was caused in the earlier DBField PR, when it seems to be this PR introducing this new typing problem.

Suddenly not being able to use some form fields for specific dbfield implementations is not something currently covered in the changelog, and should not be a problem that this PR introduces.

> No, the existing logic I left unchanged on CurrencyField is setValue(), which is not the same thing as validation.

I think you're missing the point. I'm setting a value into a form (e.g the string "this is invalid") The value is invalid - not just the wrong type like the cast to int in `DBInt::setValue()` deals with... it's just flat out invalid. That should result in a validation error message.

So either getting the validation _correct_ is in scope and this scenario should be handled in this PR, or we're explicitly not changing the validation itself, just where the code for that validation lives, in which case fixing this scenario is not in scope.

> along the way it will have also added additional fairly basic validation, such as checking string lengths of all textfields and that enum style fields e.g. Dropdowns are only allowed to select from a set of values

This sounds to me like making "this is not valid" throw a validation exception for `CurrencyField` is definitely in scope since it's very basic validation.

emteknetnz commented:
> Suddenly not being able to use some form fields for specific dbfield implementations is not something currently covered in the changelog, and should not be a problem that this PR introduces.

I don't think that's quite right, the issue in the frameworktest with Varchar/Int was that the source values of the FormField has int values, whereas they needed a string values since they were save to a Varchar column

The DBField work did essentially add a level of strong typing to the DBField layer. Strong typing enforces data integrity and reduces bugs. Are we now saying that's no longer something that's desirable?

Adding some extra words in the changelog to call this out is no problem if that's what you'd like

> CurrencyField::setValue()

There's quite a bit of existing of logic to "gracefully handle" all sorts of stuff via setValue(). This isn't validation, its data alteration.

I don't want to touch the data alteration stuff because I'm trying to limit the scope of this PR to "convert custom validation logic to FieldValidators" and "add FieldValidators where it's obvious so there is now some validation".  However what you're asking me to do is essentially change existing alteration rules.

Validation and alteration behave differently when entering an invalid value:

Alteration:
- the invalid value is reset to a default value
- there is no validation message
- does not block saving the DataObject

Validation:
- the invalid value is not reset to a default value
- there is a validation message
- will block saving the DataObject

I'm saying that changing existing alteration logic is out of scope for this PR

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -2,29 +2,114 @@
+        ?callable $converter = null,
+    ) {
+        // Convert Y-m-d args to timestamps
+        // Intermiediate variables are used to prevent "must not be accessed before initialization" PHP errors
+        // when reading properties in the constructor
```
Just set the actual properties to `null` initially instead.

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -2,29 +2,114 @@
+    /**
+     * Maximum value as a unix timestamp
+     */
+    private ?int $maxTimestamp;
```
```suggestion
    private ?string $minValue = null;

    /**
     * Minimum value as a unix timestamp
     */
    private ?int $minTimestamp = null;

    /**
     * Maximum value as a date string
     */
    private ?string $maxValue = null;

    /**
     * Maximum value as a unix timestamp
     */
    private ?int $maxTimestamp = null;
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -38,4 +123,17 @@ protected function getMessage(): string
+    /**
+     * Parse a date string into a unix timestamp using the format specified by getFormat()
+     */
+    private function dateToTimestamp(string $date): int
```
The value is `mixed` in the constructor - if a non-string is passed in, this will give a `TypeError`.

We probably should subclass the string field validator and get it to do its validation before we get to here, if this has to be a string.

emteknetnz replied:
The call

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -38,4 +123,17 @@ protected function getMessage(): string
+        // Not using symfony/validator because it was allowing d-m-Y format strings
+        $date = date_parse_from_format($this->getFormat(), $date);
+        if ($date === false || $date['error_count'] > 0 || $date['warning_count'] > 0) {
+            throw new InvalidArgumentException('Invalid date');
```
Throwing an exception is excessive here - just return `false` instead. You can give a `int|false` return type. Especially given the exceptionally broad catch which goes beyond catching this specific exception.

If there's more going on here than us throwing and catching this one exception, please add an in-code comment around the try/catch explaining what other exceptions and errors we're expecting.

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -2,29 +2,114 @@
+            }
+            $message = _t(
+                __CLASS__ . '.TOOSMALL',
+                'Value cannot be less than {minValue}',
```
I prefer the "newer" and "older" language `DateField::validate()` used. Dates are not "more" or "less".

That's fairly subjective so I won't hold merging for it, but it's probably better to err towards the existing language in cases like this IMO.

emteknetnz replied:
I've swapped out the words more/less for newer/older, though I've retained the short message, because this will also be used in a programmatic context

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -61,12 +73,10 @@ private function getFieldValidators(): array
         }
         /** @var FieldValidationInterface|Configurable $this */
-        $name = $this->getName();
+        // Use a default name if the field doesn't have one rather than throwing an exception
+        // because some fields may not have a name, i.e. CompositeField
+        $name = $this->getName() ?: _t(__CLASS__ . '.UNNAMEDFIELD', 'Unnamed field');
```
`CompositeField` does have a name - it generates one at runtime when you call `getName()` based on its children.
https://github.com/silverstripe/silverstripe-framework/blob/0255473ed8249ed01313eb9b7e5a829b2e399284/src/Forms/CompositeField.php#L123-L142

Were you testing this on an empty `CompositeField` perhaps? If that's the only outlier, we should probably either:
1. update the `CompositeField::getName()`  method so it has a name even when there's no children, or
1. override `CompositeField::getFieldValidators()` to return an empty array if there's no name (if there's no name there's no children and if there's no children there's nothing to validate anyway)

emteknetnz replied:
Have removed the fallback default name, will just use whatever $this->getName() returns, which is always a string

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -122,15 +132,30 @@ private function getClassesFromConfig(): array
+
+    /**
+     * Dedupe any cases where a subclass defines the same FieldValidator
+     * his config can happens when a subclass defines a FieldValidator that was already defined
```
```suggestion
     * This config can happen when a subclass defines a FieldValidator that was already defined
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -122,15 +132,30 @@ private function getClassesFromConfig(): array
+     */
+    private function makeUnique(array $argCalls): array
+    {
+        // there may be multiple null values, which we need to retain so make them unqiue first
```
```suggestion
        // there may be multiple null values, which we need to retain so make them unique first
```

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -122,15 +132,30 @@ private function getClassesFromConfig(): array
+            return is_null($v) ? '{null}-' . $n++ : $v;
+        }, $argCalls);
+        $argCalls = array_unique($argCalls);
+        $argCalls = array_map(fn($v) => str_contains($v, '{null}-') ? null : $v, $argCalls);
```
I personally hate both `array_map` and single letter variables because both are detrimental to readability, but I'll let it slide here for the sake of moving forward.

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -122,15 +132,30 @@ private function getClassesFromConfig(): array
+     */
+    private function makeUnique(array $argCalls): array
+    {
+        // there may be multiple null values, which we need to retain so make them unqiue first
```
What happens if the parent class gives null, but the child class wants to override that null with a specific method? Or the parent class has a specific method but the child class wants to override that with null?

I'll point out again that this isn't a problem at all if we update the injector to allow named arguments, because we can then just use key/value pairs here.

GuySartorelli replied:
Actually we can probably use key/value pairs _anyway_ in the config, but then only take the values when we fetch the config. That will dedupe automagically through config merging.

emteknetnz replied:
key/values work well, even though they're not named args. Have updated config to use them

GuySartorelli commented on src/Core/Validation/FieldValidation/MultiOptionFieldValidator.php:
```diff
@@ -12,19 +11,20 @@
      */
     public function __construct(string $name, mixed $value, array $options)
     {
-        if (!is_iterable($value) && !is_null($value)) {
-            throw new InvalidArgumentException('Value must be iterable');
-        }
```
What scenario prompted this? I'd like to try it in my browser to see what the UX is for the change.

emteknetnz replied:
It's not intended to be shown in UX. If you're passing a non-iterable in here then the solution would be to update the $field_validators config for the FormField, of the method that's called

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -429,7 +433,7 @@ public function getTemplateHelper()
-    public function getName()
+    public function getName(): string
     {
         return $this->name;
```
The property isn't typed, so can be null (or just about anything else really).
We probably need to type the property at the same time, as well as the param in the constructor.

emteknetnz replied:
Updated

Note that I'm not doing strong typing in this PR, though I have a follow up card to do this https://github.com/silverstripe/silverstripe-framework/issues/11462

emteknetnz replied:
Actually there's a load of places, particularly unit tests, though also regular code, that sets null for name. What I've opted for to keep this PR moving is to just cast to string in this method for now, and keep the dynamic type elsewhere, and look to update in the strong typing PR

GuySartorelli commented on src/Forms/ConfirmedPasswordField.php:
```diff
@@ -420,155 +421,141 @@ public function isSaveable()
-
-                return $this->extendValidationResult(false, $validator);
-            }
+            return $result;
```
In the past this was updatable with an extension hook, which we should retain.

emteknetnz replied:
There's now a central 'updateValidate' extension hook that lives in FormField::validate()

GuySartorelli replied:
Right, but this specific return happens without calling `parent::validate()`. So we need to explicitly include the `extend()` call here.

GuySartorelli commented on src/Forms/ConfirmedPasswordField.php:
```diff
@@ -420,155 +421,141 @@ public function isSaveable()
-                "validation"
-            );
-
-            return $this->extendValidationResult(false, $validator);
```
This and the other returns should be kept, as they stop subsequent validation from being run if we already hit an error.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/ConfirmedPasswordField.php:
```diff
@@ -420,155 +421,141 @@ public function isSaveable()
-
-                return $this->extendValidationResult(false, $validator);
-            }
+            return $result;
```
```suggestion
            return ValidationResult::create();
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/ConfirmedPasswordField.php:
```diff
@@ -420,155 +421,141 @@ public function isSaveable()
     {
-        $name = $this->name;
-
+        $result = ValidationResult::create();
```
Don't create one here that we're not even using except in the one case. Isntead, just `return parent::validate()` at the end.
```suggestion
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/ConfirmedPasswordField.php:
```diff
@@ -420,155 +421,141 @@ public function isSaveable()
-            $limitRegex = '/^.' . $limit . '$/';
-            if (!empty($value) && !preg_match($limitRegex ?? '', $value ?? '')) {
-                $validator->validationError(
+        $this->beforeExtending('updateValidate', function () use ($result) {
```
```suggestion
        $this->beforeExtending('updateValidate', function (ValidationResult $result) {
```

GuySartorelli replied:
Please also make this change to any other `beforeExtending()` calls you've added.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/CurrencyField.php:
```diff
@@ -54,21 +55,22 @@ public function performReadonlyTransformation()
-        }
-
-        return $this->extendValidationResult($result, $validator);
+        $result = ValidationResult::create();
+        $this->beforeExtending('updateValidate', function () use ($result) {
```
```suggestion
        $this->beforeExtending('updateValidate', function (ValidationResult $result) {
```
We should be updating the result that's passed into the extension hook.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/CurrencyField.php:
```diff
@@ -54,21 +55,22 @@ public function performReadonlyTransformation()
+                $result = false;
+            }
+        });
+        return $result->combineAnd(parent::validate());
```
```suggestion
        return parent::validate();
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/DateField.php:
```diff
@@ -536,13 +453,12 @@ protected function frontendToInternal($date)
      * @return string The formatted date, or null if not a valid date
      */
-    protected function internalToFrontend($date)
+    public function internalToFrontend(mixed $value): ?string
```
We probably shouldn't change named parameters without good cause - `$date` is a perfectly good name for this parameter.
Every change to the API adds upgrade pain so we shouldn't be making unnecessary changes.
I won't block merging for this, it's more of a strong recommendation.
```suggestion
    public function internalToFrontend(mixed $date): ?string
```

emteknetnz replied:
I've added an interfaces for the 'converter' functionality which has this

```php
interface FieldValidatorConverterInterface
{
    public function internalToFrontend(mixed $value): ?string;
```

So I think in this case we want to change the param name

GuySartorelli commented on src/Forms/EmailField.php:
```diff
@@ -2,42 +2,24 @@
-        $message = _t('SilverStripe\\Forms\\EmailField.VALIDATION', 'Please enter an email address');
-        $result = ConstraintValidator::validate(
-            $this->value,
-            new EmailConstraint(message: $message, mode: EmailConstraint::VALIDATION_MODE_STRICT),
```
This validation used `VALIDATION_MODE_STRICT` - the email field validator does not. We should use it here.
See https://symfony.com/doc/current/reference/constraints/Email.html#mode

emteknetnz replied:
Use it here? You mean update EmailFieldValidator to use it?

GuySartorelli replied:
Either way is fine, so long as when you validate an `EmailField` is uses strict mode by default.

emteknetnz replied:
Have updated EmailFieldValidator

GuySartorelli commented on src/Forms/FieldsValidator.php:
```diff
@@ -13,7 +13,9 @@ public function php($data): bool
 
         foreach ($fields as $field) {
-            $valid = ($field->validate($this) && $valid);
+            $result = $field->validate();
+            $valid = $result->isValid() && $valid;
+            $this->result->combineAnd($result);
         }
 
         return $valid;
```
```suggestion
            $this->result->combineAnd($field->validate());
        }
        return $this->result->isValid();
```
And remove the `$valid = true` at the beginning. Much more succinct and doesn't have the unnecessary variable.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -1222,32 +1234,30 @@ public function Type()
     {
-        Deprecation::notice('5.4.0', 'Use extend() directly instead');
-        $this->extend('updateValidationResult', $result, $validator);
+        if (count(func_get_args()) !== 0) {
+            // In prior versions, FormField::validate() had a method signature of
+            // validate(Validator $validator): bool
+            // Old code may still call this method this without throwing an error, however the
+            // return type is now ValidationResult which may be mistakenly interpreted as truthy
+            // While we cannot stop people from interpreting the return value as a boolean, we can
+            // at least throw an exception so they might notice something is wrong.
+            $message = 'validate() does not accept any arguments.'
+            . ' Note that the return type is ValidationResult, not boolean.';
+            throw new InvalidArgumentException($message);
+        }
```
This feels like we're doing something in CMS 6 that should be covered by a deprecation in 5 with relevant changelog information. I won't block merging for this because I can't see us agreeing about it, but IMO this is adding tech debt.

emteknetnz replied:
Removed. Have added CMS 5 deprecation notice, 5.4.0 changelog, and 6.0.0 changelog notes

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -1222,32 +1234,30 @@ public function Type()
     {
-        Deprecation::notice('5.4.0', 'Use extend() directly instead');
-        $this->extend('updateValidationResult', $result, $validator);
+        if (count(func_get_args()) !== 0) {
+            // In prior versions, FormField::validate() had a method signature of
+            // validate(Validator $validator): bool
+            // Old code may still call this method this without throwing an error, however the
+            // return type is now ValidationResult which may be mistakenly interpreted as truthy
+            // While we cannot stop people from interpreting the return value as a boolean, we can
+            // at least throw an exception so they might notice something is wrong.
+            $message = 'validate() does not accept any arguments.'
+            . ' Note that the return type is ValidationResult, not boolean.';
+            throw new InvalidArgumentException($message);
+        }
```
This feels like we're doing something in CMS 6 that should be covered by a deprecation in 5 with relevant changelog information. I won't block merging for this because I can't see us agreeing about it, but IMO this is adding tech debt.

emteknetnz replied:
Duplicate comment

GuySartorelli commented on src/Forms/FormField.php:
```diff
@@ -1222,32 +1234,30 @@ public function Type()
     {
-        Deprecation::notice('5.4.0', 'Use extend() directly instead');
-        $this->extend('updateValidationResult', $result, $validator);
+        if (count(func_get_args()) !== 0) {
```
```suggestion
        if (func_num_args() !== 0) {
```
If you do insist on keeping this, the above change is cleaner.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/ListboxField.php:
```diff
@@ -165,6 +166,17 @@ public function getDisabledItems()
 
+    public function getValueForValidation(): mixed
+    {
+        // Unlike other MultiSelectField's, ListboxField allows setting a single value
+        // Wrap single values in an array so that the validation logic can treat it as a multi-select
```
All `MultiSelectField` allows a single value, don't they? There's nothing that requires 2+ values to my knowledge.

emteknetnz replied:
Correct, though it will be a single value wrapped in an array.  Easier than having to account for a value that may or may not be in an array

GuySartorelli commented on src/Forms/LookupField.php:
```diff
@@ -15,6 +16,10 @@
  */
 class LookupField extends MultiSelectField
 {
+    private static $field_validators = [
+        MultiOptionFieldValidator::class => null,
+    ];
```
This is a read-only field, it probably shouldn't get validated.

emteknetnz replied:
Yes, the `=> null` is disabling the FieldValidator

GuySartorelli replied:
Oh yeah lol 👍

GuySartorelli commented on src/Forms/MultiSelectField.php:
```diff
@@ -72,11 +78,28 @@ public function setValue($value, $obj = null)
+    {
+        // When no value is selected and a regular edit form is submitted,
+        // usually an empty string will be POSTed e.g. MyField[]: ""
+        if ($value === [""]) {
```
```suggestion
        if ($value === ['']) {
```
Subjective stylistic change in this case, but we do tend to use the safer `''` whenever we're not injecting variables into strings.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/MultiSelectField.php:
```diff
@@ -72,11 +78,28 @@ public function setValue($value, $obj = null)
+    public function setSubmittedValue($value, $data = null)
+    {
+        // When no value is selected and a regular edit form is submitted,
+        // usually an empty string will be POSTed e.g. MyField[]: ""
```
```suggestion
        // usually an empty string will be POSTed e.g. `MyField[]: ""`
```
Just making it a little clearer that that's all part of the same code example.

emteknetnz replied:
This was updated

GuySartorelli commented on src/Forms/SingleSelectField.php:
```diff
@@ -50,6 +47,23 @@ public function getSchemaDataDefaults()
         return $data;
     }
 
+    public function setSubmittedValue($value, $data = null)
+    {
+        if (is_numeric($value) && is_string($value) && $this->getAllSourceValuesAreIntType()) {
+            $value = (int) $value;
+        }
+        parent::setSubmittedValue($value, $data);
+    }
+
```
I can't test this right now because the PRs need to be rebased for sink to dev/build with them. So this is just a reminder to me to test this as I think it'll cause new validation errors with the following code:

```php
private static array $db = [
    'MyVarchar' => 'Varchar'
];

public function getCMSFields()
{
    $fields = parent::getCMSFields();
    $fields->replaceField('MyVarchar', DropdownField::create('MyVarchar', null, ['one', 'two', 'three']));
    return $fields;
}
```

GuySartorelli replied:
Yup, can confirm this now gives a "Must be a string" validation error message which a) makes no sense for a user who's just selecting an item from a dropdown and b) didn't used to be a problem. int or string keys should work perfectly fine.

emteknetnz replied:
Have changed the "to int" casting in SingleSelectField + MultiSelectField to happen in getValueForValidation(), previously it was in setSubmittedValue().

This means value that stored internally in the FormField remains as a numeric string, rather than an int.  When it gets saved in DBVarchar there's not no validation error

DBInt will cast numeric strings to int, so things will continue work if we saved to a DBInt instead of a DBVarchar

emteknetnz replied:
I've also updated this behaviour for NumericField by largely rewritting the cast() method.  Idea is that numeric strings are used for the most part, though cast to int/float for getValueForValidation()

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -48,6 +55,14 @@ class NumericField extends TextField
 
+    public function __construct($name, $title = null, $value = null, $maxLength = null, $form = null)
+    {
+        // This constructor has a default value of null for the $value param, as opposed to the parent
+        // TextField constructor which has a default value of blank string. This is done to prevent passing
+        // a blank string to the NumericFieldValidator which which will cause a validation error.
```
Form fields should probably be robust to string/numeric conversions since field submission values will usually come back as strings and need converting anyway.

Actually I notice now that the `cast()` method used to set empty string values to `null`. It should probably just continue doing that instead of having an empty constructor.

emteknetnz replied:
> Form fields should probably be robust to string/numeric conversions

cast() method is already doing this, at least for numeric fields.  If we do decide to ditch the strong typing here, correct place to do it would be at the DBField level, not the FormField level, I think

Discussion around `null` casting on empty string can be had here https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1851171579

GuySartorelli replied:
>> Form fields should probably be robust to string/numeric conversions
>
> cast() method is already doing this

Except it isn't in this case, which is exactly my point.

emteknetnz replied:
It is casting numeric strings already?  Any non-zero length string is being cast to either (int) or (float). Code for NumericField::cast() is

```php
    protected function cast($value)
    {
        if (!is_string($value) || strlen($value) === 0) {
            return $value;
        }
        if ($this->getScale() === 0) {
            return (int) $value;
        }
        return (float) $value;
    }
```

GuySartorelli replied:
The latest changes to the `cast()` method and how it deals with null have made this conversation unnecessary 👍

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -156,21 +175,30 @@ public function setValue($value, $data = null)
         return $this;
     }
 
+    public function getValueForValidation(): mixed
+    {
+        $value = $this->value;
+        if (is_numeric($value) && is_string($value)) {
+            return $this->cast($value);
+        }
+        return $value;
+    }
+
     /**
      * Helper to cast non-localised strings to their native type
      *
      * @param string $value
-     * @return float|int
+     * @return mixed
      */
     protected function cast($value)
     {
-        if (strlen($value ?? '') === 0) {
-            return null;
+        if (!is_string($value) || strlen($value) === 0) {
+            return $value;
```
Why do we not continue treating an empty string as null?

emteknetnz replied:
So that it gets caught in validation later instead of being altered.

GuySartorelli replied:
But is an empty string not just effectively a "I don't want a value for this"? It seems weird to be throwing validation errors for it suddenly. We're casting numeric strings to `float` and `int` so we should also cast empty strings to `null`.

emteknetnz replied:
Updated to return null when $value is blank string

emteknetnz replied:
Actually have rewritten method

GuySartorelli commented on src/Forms/UrlField.php:
```diff
@@ -30,26 +31,28 @@ public function Type()
-        return $this->extendValidationResult($isValid, $validator);
+        $result = ValidationResult::create();
+        $this->beforeExtending('updateValidate', function () use ($result) {
+            $allowedProtocols = $this->getAllowedProtocols();
```
We have a url field validator - really it should be updated to have all of this functionality. I brought this up before, and you created a separate card for it so I don't expect you to change it now - just pointing out it's silly we can't use it because it doesn't have the functionality it should have.

emteknetnz replied:
There's already a card in the backlog for this - https://github.com/silverstripe/silverstripe-framework/issues/11445

GuySartorelli replied:
I know, it just feels silly to not have the functionality there already - it feels like you're doing more work in the long run.

GuySartorelli reviewed (changes requested):
I can't actually test this right now because the PRs need to be rebased before sink will be happy to dev/build, so this review is incomplete.

GuySartorelli commented on src/Core/Validation/FieldValidation/FieldValidationTrait.php:
```diff
@@ -21,13 +21,16 @@ trait FieldValidationTrait
      *    additional arg and call $field->getSomething() to get a value for the second additional arg
+     *    Keys are used to speicify the arg name, which is done to prevents duplicate
+     *    args being add to config when a subclass defines the same FieldValidator as a parent class.
+     *    Note that keys are not named args, they are simply arbitary keys
```
That note is probably not useful (people using arbitrary keys would get confusing and could result in not overriding the correct values).
```suggestion
```

Alternatively if you really want this here, we should promote best practice
```suggestion
     *    Note that keys are not named args, they are simply arbitary keys - but best practice is for the keys to match constructor argument names.
```
(if it's not the constructor just put the relevant method name instead)

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/ListboxField.php:
```diff
@@ -165,6 +166,17 @@ public function getDisabledItems()
 
+    public function getValueForValidation(): mixed
+    {
+        // Unlike other MultiSelectField's, ListboxField allows setting a single value
+        // Wrap single values in an array so that the validation logic can treat it as a multi-select
```
```suggestion
        // Unlike other MultiSelectField's, ListboxField allows setting a single value without wrapping it in an array.
        // Ensure values are wrapped in an array here so that the validation logic can treat it as a multi-select
```
This resolves the confusion I had in https://github.com/silverstripe/silverstripe-framework/pull/11449/files#r1851040162

emteknetnz replied:
Updated

GuySartorelli reviewed (changes requested):
Found while testing:
- https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1851058937
- If I have a `TextField` with a `DBInt` leaving it blank results in "Must be an integer" error even though it's not a required field. Setting numberic values works as expected.
- Setting a value of `9999999999999999999999999999999999999999999999999999999999999` for a `TextField` with `DBInt` gives the error "Must be an integer"
- Setting a value of `9999999999999999999999999999999999999999999999999999999999999` for a `NumericField` with `DBInt` gives the error "Must be numeric"
- Setting `maxLength` on a `NumericField` limits the length in the UI, but submitting a value that is too long does _not_ result in a validation error. (doing the same with `TextField` _does_ result in a validation error)
- Using `TextField` with a `DBDate` if I set the value `2024-01-24/124` I get a simple "error" toast (no validation error). In the network tab I see a 500 error with 'Uncaught InvalidArgumentException: Invalid date: '. Use y-MM-dd to prevent this error."
    - If I input `01-24-2025` I also get that exception but with a different message: "Invalid date: '01-24-2025'. Use y-MM-dd to prevent this error"
- I get similar errors as above using `TextField` with `DBDatetime` and with `DBTime` and entering invalid values.
- With `MoneyField` if I enter too many characters the validation message appears at the top of the page, instead of under the relevant form field.
- With `MoneyField` if I enter an invalid currency amount the validation message appears at the top of the page, instead of under the relevant form field.
- With `MoneyField` if I call `setAllowedCurrencies()` and set only one value, and there was an old value in the currency field before hand, saving the page results in a "Currency abc is not in the list of allowed currencies" validation error (and there's no way to change the value from the form).

Some of the errors above are likely covered under https://github.com/silverstripe/silverstripe-framework/issues/11453 - if that's the case please just list which those are and we can deal with them in that other card.


## Disabled and readonly fields

Users can't edit values in disabled or readonly fields, but those still get validated. Should we disable validation for readonly and disabled fields?

emteknetnz commented:
> DropdownField with int based source + DBVarchar

Have updated SingleSelectField and MultiSelectField to do casting string to int in use getValueForValidation() instead of setSubmittedValue(). This means that the internal value of the field stays and a numeric string

> If I have a TextField with a DBInt leaving it blank results in "Must be an integer" error even though it's not a required field. Setting numberic values works as expected.

Have update FormField::setSubmittedValue() to alter blank strings to null

> Setting a value of 9999999999999999999999999999999999999999999999999999999999999 for a TextField with DBInt gives the error "Must be an integer"
> Setting a value of 9999999999999999999999999999999999999999999999999999999999999 for a NumericField with DBInt gives the error "Must be numeric"

Note: in the PRs current state they now give "Must be an integer" validation messages

The validation message comes from IntFieldValidator, which TextField and NumericField do not have, though DBInt does. The issue here is that it's above the max size of a signed 64 bit integer which is 9223372036854775807. DBInt has logic in setValue() to cast numeric string to int, however in the case of huge numbers it cannot, so it retains the value a numeric string, hence why it fails validation. Instead of failing to casting if we throw an InvalidArgumentException we just end up with a server error toast which is bad, and, throwing ValidationException in DBInt::setValue() is just wrong. The alternative would be to "truncate" the value to the max PHP int size, but that just defeats the whole purpose of all this validation work. So I don't really know what we can really do here to improve UX. While the message isn't great, it is enforcing data integrity by preventing saving a value to the database which cannot be stored.

> Setting maxLength on a NumericField limits the length in the UI, but submitting a value that is too long does not result in a validation error. (doing the same with TextField does result in a validation error)

Have added created NumericNonStringFieldValidator and moved the 'is_string() && is_numeric() to it, and gotten DBInt / DBDouble to extend that. NumericField uses the NumericStringFieldValidator so it does allow string values

> Using TextField with a DBDate if I set the value 2024-01-24/124 I get a simple "error" toast (no validation error). In the network tab I see a 500 error with 'Uncaught InvalidArgumentException: Invalid date: '. Use y-MM-dd to prevent this error."
> If I input 01-24-2025 I also get that exception but with a different message: "Invalid date: '01-24-2025'. Use y-MM-dd to prevent this error"
> I get similar errors as above using TextField with DBDatetime and with DBTime and entering invalid values.

These are all out of scope as this is existing behaviour. The exceptions are coming from DBDate/DBTime::setValue()

> With MoneyField if I enter too many characters the validation message appears at the top of the page, instead of under the relevant form field.
> With MoneyField if I enter an invalid currency amount the validation message appears at the top of the page, instead of under the relevant form field.

Have updated FieldValidationTrait to remove the subfield portion of the name so that the validation message now shows on the composite field

> With MoneyField if I call setAllowedCurrencies() and set only one value, and there was an old value in the currency field before hand, saving the page results in a "Currency abc is not in the list of allowed currencies" validation error (and there's no way to change the value from the form).

MoneyField is weird to begin with in that the currency field can either be a DropDownField (two or more allowed currencies) a HiddenField (one allowed currency) or a TextField (zero allowed currencies). The issue you had is that when there's only one allowed currency, it becomes a HiddenField. I've updated setValue() to swap out the currency field to a TextField if the currency being set is not in the allowed value list. This isn't ideal and I've rather do this in the MoneyField constructor, but when the form is rendering the value is set via setValue(), rather than as a contructor argument.

> Users can't edit values in disabled or readonly fields, but those still get validated. Should we disable validation for readonly and disabled fields?

Out of scope as these are existing issues, e.g.
- $fields->replaceField('MyVarchar', UrlField::create('MyVarchar'));
- Fill in value as http://example.com and save form
- $fields->replaceField('MyVarchar', UrlField::create('MyVarchar')->setAllowedProtocols(['https'])->setReadonly(true));
- Save form - validation error you cannot alter. Same issue with setDisabled(true)

emteknetnz commented on src/Forms/NumericField.php:
```diff
@@ -104,15 +121,26 @@ protected function getNumberType()
         return trim(str_replace(["\u{00A0}", "\u{202F}", "\u{2009}"], ' ', $value ?? ''));
     }
 
+    public function setValue($value, $data = null)
```
This has been moved up from lower down in the file

GuySartorelli commented on src/Forms/MoneyField.php:
```diff
@@ -321,32 +338,30 @@ public function getLocale()
-        return $this->extendValidationResult($result, $validator);
+    public function validate(): ValidationResult
+    {
+        $result = ValidationResult::create();
+        $this->beforeExtending('updateValidate', function () use ($result) {
```
```suggestion
        $this->beforeExtending('updateValidate', function (ValidationResult $result) {
```
I've asked this in a previous review: Please double check all of the changed fields to ensure you are doing it this way instead, so the same result passed into the extension hook is used.
This is the only one I found this time around but I may have missed one.

emteknetnz replied:
I think this was the only one to slip through, have updated

GuySartorelli commented on src/Forms/MoneyField.php:
```diff
@@ -321,32 +338,30 @@ public function getLocale()
+                );
+            }
+        });
+        return $result->combineAnd(parent::validate());
```
```suggestion
        return parent::validate();
```

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/SingleSelectField.php:
```diff
@@ -50,6 +47,17 @@ public function getSchemaDataDefaults()
+        $value = $this->getValue();
+        if ($value === '' && $this->getHasEmptyDefault()) {
+            return null;
+        } elseif (is_string($value) && is_numeric($value) && $this->getAllSourceValuesAreIntType()) {
+            return (int) $value;
+        }
```
I think this is an incomplete solution. If I have

```php
DropdownField::create('MyField', null, ['one', 'two' => 'two', 'three']);
// or
DropdownField::create('MyField', null, [1 => 'one', 'two' => 'two',  '3' => 'three']);
```

The integer keys are not being converted to `int` so I get a "Not an allowed value" error if selecting "one" from the dropdown. Or at least I assume that's what's causing the error.

This method probably needs to check against the keys in `$this->getSource()` and only cast to int if it's an int in the source, or something along those lines.

Probably also needs an inline comment explaining why the cast is happening - without the context of this PR it'll be pretty hard to just look at it and understand why it's doing what it's doing.

emteknetnz replied:
Have redone solution, remove SelectField::getAllSourceValuesAreIntType() and created SelectField::castSubmittedValue() instead

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -149,28 +178,36 @@ public function Value()
+    protected function cast(mixed $value): mixed
     {
-        if (strlen($value ?? '') === 0) {
+        if ($value === false) {
+            return false;
+        }
```
What situation is this happening in? Why `false` explicitly and not `is_boolean()`?

emteknetnz replied:
$value will be `false` when the localisation formatter fails to parse the value in setSubmittedValue()

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -149,28 +178,36 @@ public function Value()
     }
 
     /**
      * Helper to cast non-localised strings to their native type
```
This doesn't seem to be what's happening anymore - we're casting to string after casting away from string, apparently.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -149,28 +178,36 @@ public function Value()
+        // e.g. trying to cast an array to a string
+        try {
+            if ($this->getScale() === 0) {
+                return (string) (int) $value;
+            }
+            return (string) (float) $value;
```
This feels wrong. What is the specific intention here? Casting to int and then to string seems pointless? Is this try/catch a non-standard conditional that could be replaced with a `is_numeric()` check?

emteknetnz replied:
$value could be anything, e.g. float, int, numeric string.  There was existing logic to chop of decimal places if $this->getScale() is 0, hence the int cast.  

I've updated the docblock description to say that the point of cast() is to get a numeric string and repect getScale()

Have changed try/catch to is_numeric()

GuySartorelli replied:
> This feels wrong. What is the specific intention here? Casting to int and then to string seems pointless?

This part still applies. I still don't understand what the purpose of this is. Perhaps it just needs an inline code comment explaining it? But you might also need to explain it to me in person so I really get it. It's definitely weird and it's not clear what the intention is.

emteknetnz replied:
I've retained the logic though refactored the class and added a bunch of inline comments explaining all the things

GuySartorelli replied:
Thank you, that's much clearer - and splitting up the casts makes it look more sensible even if the end result is the same.

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -232,7 +244,16 @@ public function getSchemaValidation()
     {
-        return $this->cast($this->value);
+        $value = $this->getValue();
+        if ($value === '') {
+            return null;
```
This already happens in `cast` - we probably don't need to do it here?

emteknetnz replied:
Removed

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -232,7 +244,16 @@ public function getSchemaValidation()
+        $value = $this->getValue();
+        if ($value === '') {
+            return null;
+        } elseif ($value === false) {
+            return 0;
+        }
```
1. When would the value be explicitly `false`?
1. Why do we return `0` here, but cast returns `false`?

emteknetnz replied:
false $value happens when the localisation formatter fails to parse the in setSubmittedValue()

dataValue() is passed to the DBField, we do not want to be settings the DBField value to false. 0 is appropriate.

GuySartorelli replied:
We shouldn't be passing the _wrong_ value into `DBInt`. If someone sets an invalid value to the form field, we shouldn't be replacing whatever value used to be stored in the database with `0`

 Let it return false from here, and the validation logic in `DBInt` will pick it up.

emteknetnz replied:
That's true. Won't necessarily be DBInt as it could be any DBField e.g. DBVarchar. However NumericField::validate() will prevent code from even reaching this point so it doesn't really matter. I've expanded on the inline comment to explain this. In hindsight the only reason I was returning zero here was to get an existing unit test to pass, I've now updated the unit test to expect false instead of 0

GuySartorelli replied:
> However NumericField::validate() will prevent code from even reaching this point so it doesn't really matter.

It won't in a scenario where someone is calling methods directly - and in that case setting `0` instead of letting `DBField` validation kick in would have been inappropriate.

GuySartorelli commented on src/Forms/NumericField.php:
```diff
@@ -149,28 +178,36 @@ public function Value()
-        $this->value = $this->cast($value);
-        return $this;
+        $value = $this->value;
+        if (is_numeric($value) && is_string($value)) {
+            return $this->cast($value);
+        }
```
Why only call `cast()` if that condition is true?

emteknetnz replied:
Have removed, we no longer need to cast since the FieldValidator now accepts numeric strings

GuySartorelli commented on src/Core/Validation/FieldValidation/IntFieldValidator.php:
```diff
@@ -8,7 +8,7 @@
  * Validates that a value is a 32-bit signed integer
  */
-class IntFieldValidator extends NumericFieldValidator
+class IntFieldValidator extends NumericNonStringFieldValidator
```
Remove the `use` statement

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DecimalFieldValidator.php:
```diff
@@ -3,7 +3,7 @@
 
 use SilverStripe\Core\Validation\ValidationResult;
-use SilverStripe\Core\Validation\FieldValidation\NumericFieldValidator;
+use SilverStripe\Core\Validation\FieldValidation\NumericNonStringFieldValidator;
```
```suggestion
```
Shouldn't be needed since it's in the same namespace

emteknetnz replied:
Updated

GuySartorelli commented on src/Core/Validation/FieldValidation/DateFieldValidator.php:
```diff
@@ -2,29 +2,109 @@
+use Exception;
+use InvalidArgumentException;
 use SilverStripe\Core\Validation\ValidationResult;
+use SilverStripe\Core\Validation\FieldValidation\StringFieldValidator;
```
```suggestion
```
Shouldn't be needed since it's in the same namespace. Please check the other field validators for redundant `use` statements.

emteknetnz replied:
Updated

GuySartorelli commented on src/Forms/MoneyField.php:
```diff
@@ -175,8 +184,16 @@ public function setValue($value, $data = null)
             throw new InvalidArgumentException("Invalid currency format");
         }
 
+        // Check that currency is allowed, if not, swap out currency field for a text field so
+        // that user can alter the value as it will fail validation
+        $allowedCurrencies = $this->getAllowedCurrencies() ?? [];
+        $currency = $value['Currency'];
+        if ($currency && count($allowedCurrencies) && !in_array($currency, $allowedCurrencies)) {
+            $this->fieldCurrency = $this->buildCurrencyField(true);
+        }
```
This provides the option to change the currency, but the validation error message is simply "Currency USD is not in the list of allowed currencies"

Either the validation message, or `setDescription()` on the currency field, needs to include the list of allowed currencies. Otherwise content authors will be going through the full list of ISO currencies trying to set it to the only one that's actually allowed.

emteknetnz replied:
Updated

GuySartorelli reviewed (changes requested):
Still not done: https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1855592936

## Numeric field problem

`DBInt` with its default scaffolded form (`NumericField`) and a value of "99999999999999999999999999999999999999999999999999999999999999999999" now says "Must be a string" as well as "Must be numeric"
![image](https://github.com/user-attachments/assets/4d124e49-18dd-4a19-b74f-0b989ee6f1ed)

It must _not_ be a string, and it _is_ numeric so these error messages are clearly wrong.

Regarding the "Must be numeric" part of this which I mentioned last review, you said

> The issue here is that it's above the max size of a signed 64 bit integer which is 9223372036854775807. DBInt has logic in setValue() to cast numeric string to int, however in the case of huge numbers it cannot, so it retains the value a numeric string, hence why it fails validation.

One option to provide a more accurate validation message would be to check:
1. Is the original value a numeric string? (`is_numeric('99999999999999999999999999999999999999999999999999999999999999999999')` will return `true` even though it's larger than the int that can be represented by PHP)
1. Does casting the value to int result in a value different to the original? (e.g. `(int)$val == $val` will return `true` if the value can be represented as a PHP int, but `false` if it cannot)

## Still not resolved from previous review

> With MoneyField if I enter too many characters the validation message appears at the top of the page, instead of under the relevant form field.

> With MoneyField if I enter an invalid currency amount the validation message appears at the top of the page, instead of under the relevant form field.

emteknetnz commented:
> Still not done: https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1855592936

It's following the standard pattern of `$this->beforeExtending('updateValidate', function (ValidationResult $result) {` followed by `return parent::validate();`, meaning that `$this->extend('updateValidate')` is called in `FormField::validate()`.  I'm not sure what the issue is?

**Update: have changed the `return ValidateResult::create();` to `return parent::validate();` so that the extend hook is called in the parent class**

>> Numeric field problem

I've added "manual" validation to NumericField to check if `$this->getValue() === false`, which means the localisation formatter was unable to parse the number.  In the '999999999999999999999999999999999999999999' scenario I think it's simply too large i.e. bigger than a 64 bit int, so the formatter cannot parse it.

It'll now simply give a validation message of "Invalid number", which is the same message if you enter "fish".  `$this->getFormatter()->getErrorMessage()` gives non-useful message e.g. `U_ZERO_ERROR` so we cannot use that in the error message.

> With MoneyField if I enter an invalid currency amount the validation message appears at the top of the page, instead of under the relevant form field.

As noted [here](https://github.com/silverstripe/silverstripe-framework/pull/11449#issuecomment-2499203291), that was previously resolved, currently works as it should

> With MoneyField if I enter too many characters the validation message appears at the top of the page, instead of under the relevant form field.

Even though the PR has added the validation, that's an existing issue. This is currently happened because of the newly added `NumericField::validate()` logic which adds a field error where $this->getName() is `MyMoney[Amount]`.

In order for this to show in the correct location `$this->getName()` would need to be `MyMoney`.  This issue has to do with ValidationException thrown by FormFields in CompositeField's in general.  The NumericField, or any FormField in general is not aware that it's in a CompositeField. So I'd say this is out of scope.

GuySartorelli commented:
I don't understand how having the `Currency` field from a `MoneyField` display validation errors in the correct place is _in_ scope, but doing the same thing for the `Amount` field from a `MoneyField` is _out of_ scope, but I'll leave it in the interest of getting this merged.

GuySartorelli reviewed (changes requested):
Sorry, didn't start a review before adding comments.
Unresolved conversations:
- https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1859402941
- https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1859404386

emteknetnz committed:
ENH Use FieldValidators for FormField validation

GuySartorelli commented:
There's a merge conflict

emteknetnz commented:
Updated

GuySartorelli reviewed (approved):
LGTM
